rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java

```diff
@@ -315,19 +329,63 @@ public static LockWhitelist deserializeLockWhitelist(byte[] data, NetworkParamet
             throw new RuntimeException("deserializeLockWhitelist: expected an even number of addresses, but odd given");
         }
 
-        Map<Address, Coin> whitelist = new HashMap<>(serializedAddressesSize / 2);
+        Map<Address, LockWhitelistEntry> whitelist = new HashMap<>(serializedAddressesSize / 2);
         for (int i = 0; i < serializedAddressesSize; i = i + 2) {
             byte[] hash160 = rlpList.get(i).getRLPData();
             byte[] maxValueData = rlpList.get(i + 1).getRLPData();
+            Address address = new Address(parameters, hash160);
             whitelist.put(
-                new Address(parameters, hash160),
-                Coin.valueOf(safeToBigInteger(maxValueData).longValueExact())
+                    address,
+                new OneOffWhiteListEntry(address, Coin.valueOf(safeToBigInteger(maxValueData).longValueExact()))
             );
         }
         int disableBlockHeight = safeToBigInteger(rlpList.get(serializedAddressesSize).getRLPData()).intValueExact();
         return new LockWhitelist(whitelist, disableBlockHeight);
     }
 
+    public static LockWhitelist deserializeLockWhitelist(byte[] data, byte[] unlimitedLockWhiteListData, NetworkParameters parameters) {
```

Comment:
I'd probably leave the logic of "if there is unlimited data present then build the entries" to the storage side, which is the one more aware of what things are and are not stored. I'd leave these methods as simple as possible, i.e., have one method deserialize one-off entries as a list and another one deserialize unlimited entries as a list.